How many late days are you using for this assignment?  1

In [11]:
# Imports and pip installations (if needed)
import pandas as pd
from sklearn.preprocessing import MinMaxScaler

# Part 1: Load the dataset

In [3]:
# Load the given datasets
ckd_cat = pd.read_csv("chronic_kidney_disease_categorical.csv")
ckd_num = pd.read_csv("chronic_kidney_disease_numerical.csv")
# Print the data
print("Categorical dataset:")
print(ckd_cat.head())

print("\nNumerical dataset:")
print(ckd_num.head())

Categorical dataset:
   unique_id   al   su       rbc        pc         pcc          ba  htn   dm  \
0     224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes  yes   
1     992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   no   
2     308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   no   
3     450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   no   
4     881763  0.0  0.0    normal    normal  notpresent  notpresent   no   no   

  cad appet   pe  ane  Target  
0  no  good  yes   no     ckd  
1  no  good   no   no  notckd  
2  no  good  yes   no     ckd  
3  no  good   no  yes     ckd  
4  no  good   no   no  notckd  

Numerical dataset:
   unique_id   age    bp    bgr     bu   sc    sod  pot  hemo   pcv    wbcc  \
0     203694  38.0  80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0  7300.0   
1     938027  43.0  60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0  7200.0   
2     421471  37.0  60.0  111.0   35.0  0.8  135.

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer:
The dataset is collected for analysis and prediction on the health condition of individuals in relation to chronic kidney disease. 

The features are: age, blood pressure. specific gravity, albumin, sugar, red blood cells, pus cell, pus cell clumps, bacteria, blood urea, serum creatinine, sodium, potassium, hemoglobin, packed cell volume, white blood cell count, hypertension, diabetes mellitus, coronary artery disease, appetite, pedal edema, and anemia.

The label is ckd or not ckd - if the patient has/does not have chronic kidney disease. 

The features in the dataset are a mix of both. 

Categorical variables:
- sg: specific gravity
- al: albumin
- su: sugar

Numerical variables:
- age: age
- bp: blood pressure
- bgr: blood glucose random

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [4]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
numerical_duplicates = ckd_num['unique_id'].duplicated(keep=False)
number_of_numerical_duplicates = numerical_duplicates.sum()
print(f"Number of numerical duplicates in 'unique_id': {number_of_numerical_duplicates} ")
print("Duplicates: \n", numerical_duplicates)
# Drop these duplicate rows
ckd_num = ckd_num.drop_duplicates(subset=['unique_id'], keep = 'first')

# Repeat the same for categorical dataset. Print the duplicate rows and drop them

categorical_duplicates = ckd_cat['unique_id'].duplicated(keep=False)
number_of_categorical_duplicates = categorical_duplicates.sum()
print(f"Number of categorical duplicates in 'unique_id': {number_of_categorical_duplicates}")
print("Duplicates: \n", categorical_duplicates)

ckd_cat = ckd_cat.drop_duplicates(subset=['unique_id'], keep = 'first')



Number of numerical duplicates in 'unique_id': 59 
Duplicates: 
 0      False
1       True
2      False
3      False
4       True
       ...  
415    False
416    False
417    False
418    False
419    False
Name: unique_id, Length: 420, dtype: bool
Number of categorical duplicates in 'unique_id': 59
Duplicates: 
 0       True
1      False
2      False
3      False
4      False
       ...  
425     True
426     True
427     True
428     True
429     True
Name: unique_id, Length: 430, dtype: bool


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer: No, not all unique ids are present in both datasets. By finding all unique_ids in both datasets and finding the difference, we can see that there are unique_ids that are in one dataset but not the other. There are unique_ids in the categorical dataset that are not present in the numerical dataset. By printing out a missing unique_id (208034), we see that the row for this unique_id in the combined table shows that data values for the numerical table are NaN, or missing. 

In [5]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
combined_df = pd.merge(ckd_num, ckd_cat, on='unique_id', how='outer')
#Print the combined dataset
print("Combined dataset: \n", combined_df)

num_ids = set(ckd_num['unique_id'])
cat_ids = set(ckd_cat['unique_id'])

missing_in_numerical = cat_ids - num_ids
print(f"Missing 'unique_id' values in numerical: {missing_in_numerical}")

missing_in_categorical= num_ids - cat_ids
print(f"Missing 'unique_id' values in categorical: {missing_in_categorical}")


print(f"Missing row in combined table: {combined_df.loc[combined_df['unique_id'] == 208034]}")

Combined dataset: 
      unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1       103570  24.0  80.0  125.0    NaN   NaN  136.0  3.5  15.4  43.0  ...   
2       104675  65.0  80.0  215.0  133.0   2.5    NaN  NaN  13.2  41.0  ...   
3       105923  90.0  90.0  139.0   89.0   3.0  140.0  4.1  12.0  37.0  ...   
4       107297  60.0  70.0    NaN    NaN   NaN    NaN  NaN  16.4  43.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
395     995177  69.0  70.0  214.0   96.0   6.3  120.0  3.9   9.4  28.0  ...   
396     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   
397     996412  70.0  60.0  120.0   40.0   0.5  140.0  4.6  16.0  43.0  ...   
398     997629  35.0   NaN   99.0   30.0   0.5  135.0  4.9  15.4  48.0  ...   
399     997876  64.0  70.0  107.0   15.0   NaN    NaN  NaN  12.8  38.0  ...   

           pc         pcc      

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [6]:
# Calculate the percentage of rows that contain atleast one missing value
total_rows = len(combined_df)
num_rows_with_missing = combined_df.isnull().any(axis = 1).sum()

rows_missing_percentage = (num_rows_with_missing / total_rows) * 100

# Print %
print(f"Percentage of rows with at least one missing value: {rows_missing_percentage}%")
# Drop these rows from the dataset
cleaned_combined_df = combined_df.dropna()

# Print the Dataset
print("Cleaned, combined, dataset: \n", cleaned_combined_df)

Percentage of rows with at least one missing value: 61.75000000000001%
Cleaned, combined, dataset: 
      unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
6       109053  64.0  60.0  106.0   27.0   0.7  150.0  3.3  14.4  42.0  ...   
7       114717  41.0  80.0  122.0   25.0   0.8  138.0  5.0  17.1  41.0  ...   
8       118191  42.0  70.0   93.0   32.0   0.9  143.0  4.7  16.6  43.0  ...   
9       118596  74.0  60.0   88.0   50.0   0.6  147.0  3.7  17.2  53.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
390     980291  62.0  70.0  122.0   42.0   1.7  136.0  4.7  12.6  39.0  ...   
391     991031  30.0  80.0   89.0   42.0   0.5  139.0  5.0  16.7  52.0  ...   
394     994377  71.0  70.0   81.0   18.0   0.8  145.0  5.0  14.7  44.0  ...   
395     995177  69.0  70.0  214.0   96.0   6.3  120.0  3.9   9.4  28.0  ...   
396     996076  57.0  60.0  10

## Part 3.4: Sort the dataset according to the Labels

In [7]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
sorted_df = cleaned_combined_df.sort_values(by='Target').reset_index()
# Print the dataset
print("Grouped dataset: \n", sorted_df)

Grouped dataset: 
      index  unique_id   age    bp    bgr     bu    sc    sod  pot  hemo  ...  \
0        0     102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  ...   
1      103     337834  48.0  70.0  117.0   56.0   3.8  111.0  2.5  11.2  ...   
2      105     343710  61.0  80.0  173.0  148.0   3.9  135.0  5.2   7.7  ...   
3      108     349892  59.0  70.0  424.0   55.0   1.7  138.0  4.5  12.6  ...   
4      110     353025  40.0  70.0  253.0  150.0  11.9  132.0  5.6  10.9  ...   
..     ...        ...   ...   ...    ...    ...   ...    ...  ...   ...  ...   
148    171     486397  20.0  70.0  123.0   44.0   1.0  135.0  3.8  14.6  ...   
149     43     203466  22.0  60.0   97.0   18.0   1.2  138.0  4.3  13.5  ...   
150    169     482105  44.0  60.0   95.0   46.0   0.5  138.0  4.2  15.0  ...   
151     54     228849  58.0  80.0  100.0   50.0   1.2  140.0  3.5  14.0  ...   
152    396     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  ...   

           pc       

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [23]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
encoded_df = pd.get_dummies(sorted_df, columns=categorical_columns, drop_first=False)
# Print the dataset
print("Encoded dataset: \n", encoded_df)

Encoded dataset: 
      index  unique_id   age    bp    bgr     bu    sc    sod  pot  hemo  ...  \
0        0     102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  ...   
1      103     337834  48.0  70.0  117.0   56.0   3.8  111.0  2.5  11.2  ...   
2      105     343710  61.0  80.0  173.0  148.0   3.9  135.0  5.2   7.7  ...   
3      108     349892  59.0  70.0  424.0   55.0   1.7  138.0  4.5  12.6  ...   
4      110     353025  40.0  70.0  253.0  150.0  11.9  132.0  5.6  10.9  ...   
..     ...        ...   ...   ...    ...    ...   ...    ...  ...   ...  ...   
148    171     486397  20.0  70.0  123.0   44.0   1.0  135.0  3.8  14.6  ...   
149     43     203466  22.0  60.0   97.0   18.0   1.2  138.0  4.3  13.5  ...   
150    169     482105  44.0  60.0   95.0   46.0   0.5  138.0  4.2  15.0  ...   
151     54     228849  58.0  80.0  100.0   50.0   1.2  140.0  3.5  14.0  ...   
152    396     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  ...   

     cad_no  cad_yes

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: When encoding categorical variables for modeling, the approach differs significantly between binary and multi-class categories. For binary variables (e.g., "Yes" and "No"), using a single column is effective and straightforward, allowing models to interpret the variable correctly as two distinct states, often encoded as `0` and `1`. However, for multi-class categorical variables, such as a "type of condiment" that can take multiple values (e.g., "Ketchup," "Mustard," "Mayo," "Hot Sauce," "Relish"), using a single column can be problematic as it may lead to a misinterpretation of categories as ordinal, potentially skewing model predictions. Instead, one-hot encoding is the preferred method for handling such multi-class categories, as it creates separate binary columns for each category, allowing the model to treat them independently without implying any ordinal relationship. While one-hot encoding increases dimensionality and the risk of multicollinearity in certain types of models, it provides a clearer representation of categorical data that enhances model performance and interpretability.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [9]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
def remove_outliers(df, column):
    mean = df[column].mean()  
    std = df[column].std()  
    lower_bound = mean - 3 * std  
    upper_bound = mean + 3 * std 
    return df[(df[column] >= lower_bound) & (df[column] <= upper_bound)]

numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

# Apply the function to each numerical column
for column in numerical_columns:
    cleaned_combined_df = remove_outliers(cleaned_combined_df, column)

# Print the cleaned dataset
print("Dataset after removing outliers:")
print(cleaned_combined_df)
# Print the dataset

Dataset after removing outliers:
     unique_id   age    bp    bgr    bu   sc    sod  pot  hemo   pcv  ...  \
6       109053  64.0  60.0  106.0  27.0  0.7  150.0  3.3  14.4  42.0  ...   
7       114717  41.0  80.0  122.0  25.0  0.8  138.0  5.0  17.1  41.0  ...   
8       118191  42.0  70.0   93.0  32.0  0.9  143.0  4.7  16.6  43.0  ...   
9       118596  74.0  60.0   88.0  50.0  0.6  147.0  3.7  17.2  53.0  ...   
11      124304  69.0  70.0  264.0  87.0  2.7  130.0  4.0  12.5  37.0  ...   
..         ...   ...   ...    ...   ...  ...    ...  ...   ...   ...  ...   
390     980291  62.0  70.0  122.0  42.0  1.7  136.0  4.7  12.6  39.0  ...   
391     991031  30.0  80.0   89.0  42.0  0.5  139.0  5.0  16.7  52.0  ...   
394     994377  71.0  70.0   81.0  18.0  0.8  145.0  5.0  14.7  44.0  ...   
395     995177  69.0  70.0  214.0  96.0  6.3  120.0  3.9   9.4  28.0  ...   
396     996076  57.0  60.0  105.0  49.0  1.2  150.0  4.7  15.7  44.0  ...   

           pc         pcc          ba  htn

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [13]:

scaler = MinMaxScaler()


cleaned_combined_df[numerical_columns] = scaler.fit_transform(cleaned_combined_df[numerical_columns])

# Print the normalized dataset
print("Dataset after normalization:")
print(cleaned_combined_df.head())

Dataset after normalization:
    unique_id       age    bp       bgr        bu        sc       sod  \
6      109053  0.783784  0.00  0.154506  0.111111  0.037037  1.000000   
7      114717  0.472973  0.50  0.223176  0.098039  0.049383  0.600000   
8      118191  0.486486  0.25  0.098712  0.143791  0.061728  0.766667   
9      118596  0.918919  0.00  0.077253  0.261438  0.024691  0.900000   
11     124304  0.851351  0.25  0.832618  0.503268  0.283951  0.333333   

         pot      hemo       pcv  ...        pc         pcc          ba  htn  \
6   0.137931  0.663366  0.612903  ...    normal  notpresent  notpresent   no   
7   0.724138  0.930693  0.580645  ...    normal  notpresent  notpresent   no   
8   0.620690  0.881188  0.645161  ...    normal  notpresent  notpresent   no   
9   0.275862  0.940594  0.967742  ...    normal  notpresent  notpresent   no   
11  0.379310  0.475248  0.451613  ...  abnormal  notpresent  notpresent  yes   

     dm  cad appet   pe ane  Target  
6    no   no 

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: Yes, unique id is a unique identifier and does not tell us anything about the patients health.

In [14]:
# Remove unnecessary columns
columns_to_drop = ['unique_id']  # Add other column names as needed
cleaned_combined_df = cleaned_combined_df.drop(columns=columns_to_drop)

# Print the dataset after dropping unnecessary columns
print("Dataset after removing unnecessary columns: \n", cleaned_combined_df)


Dataset after removing unnecessary columns: 
           age    bp       bgr        bu        sc       sod       pot  \
6    0.783784  0.00  0.154506  0.111111  0.037037  1.000000  0.137931   
7    0.472973  0.50  0.223176  0.098039  0.049383  0.600000  0.724138   
8    0.486486  0.25  0.098712  0.143791  0.061728  0.766667  0.620690   
9    0.918919  0.00  0.077253  0.261438  0.024691  0.900000  0.275862   
11   0.851351  0.25  0.832618  0.503268  0.283951  0.333333  0.379310   
..        ...   ...       ...       ...       ...       ...       ...   
390  0.756757  0.25  0.223176  0.209150  0.160494  0.533333  0.620690   
391  0.324324  0.50  0.081545  0.209150  0.012346  0.633333  0.724138   
394  0.878378  0.25  0.047210  0.052288  0.049383  0.833333  0.724138   
395  0.851351  0.25  0.618026  0.562092  0.728395  0.000000  0.344828   
396  0.689189  0.00  0.150215  0.254902  0.098765  1.000000  0.620690   

         hemo       pcv      wbcc  ...        pc         pcc          ba  htn

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [15]:
# Export the dataframe to a new csv file
cleaned_combined_df.to_csv('cleaned_chronic_kidney_disease_data.csv', index=False)

# Export the dataframe to a new json file
cleaned_combined_df.to_json('cleaned_chronic_kidney_disease_data.json', orient='records', lines=True)

print("Data exported successfully in CSV and JSON formats.")


Data exported successfully in CSV and JSON formats.


# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

Paste here: 
|    | age      | bp    | bgr      | bu       | sc       | sod      | pot      | hemo     | pcv      | wbcc     | ...      | pc       | pcc        | ba        | htn  | dm   | cad  | appet | pe   | ane  | Target |
|----|----------|-------|----------|----------|----------|----------|----------|----------|----------|----------|----------|----------|------------|-----------|------|------|------|-------|------|------|--------|
| 6  | 0.783784 | 0.00  | 0.154506 | 0.111111 | 0.037037 | 1.000000 | 0.137931 | 0.663366 | 0.612903 | 0.306452 | ...      | normal   | notpresent | notpresent | no   | no   | no   | good  | no   | no   | notckd |
| 7  | 0.472973 | 0.50  | 0.223176 | 0.098039 | 0.049383 | 0.600000 | 0.724138 | 0.930693 | 0.580645 | 0.387097 | ...      | normal   | notpresent | notpresent | no   | no   | no   | good  | no   | no   | notckd |
| 8  | 0.486486 | 0.25  | 0.098712 | 0.143791 | 0.061728 | 0.766667 | 0.620690 | 0.881188 | 0.645161 | 0.225806 | ...      | normal   | notpresent | notpresent | no   | no   | no   | good  | no   | no   | notckd |
| 9  | 0.918919 | 0.00  | 0.077253 | 0.261438 | 0.024691 | 0.900000 | 0.275862 | 0.940594 | 0.967742 | 0.137097 | ...      | normal   | notpresent | notpresent | no   | no   | no   | good  | no   | no   | notckd |
| 11 | 0.851351 | 0.25  | 0.832618 | 0.503268 | 0.283951 | 0.333333 | 0.379310 | 0.475248 | 0.451613 | 0.427419 | ...      | abnormal | notpresent | notpresent | yes  | yes  | yes  | good  | yes  | no   | ckd    |
| 13 | 0.418919 | 0.00  | 0.167382 | 0.241830 | 0.086420 | 0.700000 | 0.689655 | 0.722772 | 0.806452 | 0.217742 | ...      | normal   | notpresent | notpresent | no   | no   | no   | good  | no   | no   | notckd |
| 14 | 0.567568 | 0.50  | 0.270386 | 0.843137 | 1.000000 | 0.400000 | 0.896552 | 0.257426 | 0.419355 | 0.153226 | ...      | abnormal | notpresent | present    | no   | yes  | no   | good  | yes  | no   | ckd    |
| 15 | 0.621622 | 0.50  | 0.236052 | 0.078431 | 0.098765 | 0.633333 | 0.586207 | 0.871287 | 0.645161 | 0.032258 | ...      | normal   | notpresent | notpresent | no   | no   | no   | good  | no   | no   | notckd |
| 16 | 0.310811 | 0.50  | 0.055794 | 0.254902 | 0.061728 | 0.633333 | 0.137931 | 0.970297 | 0.548387 | 0.451613 | ...      | normal   | notpresent | notpresent | no   | no   | no   | good  | no   | no   | notckd |
| 18 | 0.905405 | 1.00  | 0.965665 | 0.522876 | 0.641975 | 0.666667 | 0.000000 | 0.148515 | 0.225806 | 0.217742 | ...      | abnormal | present    | notpresent | yes  | yes  | yes  | poor  | no   | no   | ckd    |
| 25 | 0.837838 | 0.00  | 0.236052 | 0.202614 | 0.086420 | 0.633333 | 0.310345 | 0.960396 | 0.870968 | 0.193548 | ...      | normal   | notpresent | notpresent | no   | no   | no   | good  | no   | no   | notckd |
| 26 | 0.716216 | 0.00  | 0.184549 | 0.084967 | 0.086420 | 0.633333 | 0.206897 | 0.752475 | 1.000000 | 0.177419 | ...      | normal   | notpresent | notpresent | no   | no   | no   | good  | no   | no   | notckd |
| 32 | 0.567568 | 0.00  | 0.180258 | 0.222222 | 0.098765 | 0.733333 | 0.689655 | 0.673267 | 0.677419 | 0.411290 | ...      | normal   | notpresent | notpresent | no   | no   | no   | good  | no   | no   | notckd |
| 33 | 0.810811 | 0.25  | 0.158798 | 0.039216 | 0.086420 | 0.666667 | 0.241379 | 0.584158 | 0.612903 | 0.540323 | ...      | normal   | notpresent | notpresent | no   | no   | no   | good  | no   | no   | notckd |
| 36 | 0.524752 | 0.00  | 0.124464 | 0.183007 | 0.049383 | 0.500000 | 0.275862 | 0.524752 | 0.838710 | 0.322581 | ...      | normal   | notpresent | notpresent | no   | no   | no   | good  | no   | no   | notckd |


### Paste the markdown table here

** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

When using language models for data conversions, it is crucial to mitigate potential hallucinations—instances where the model generates inaccurate or misleading information. To do this, start by providing clear and specific instructions, including examples of the desired output format to reduce ambiguity. Limit the amount of input data to prevent overwhelming the model, and always manually review the generated outputs against the original dataset to identify inconsistencies. Additionally, setting constraints on output formats can help maintain consistency, while an iterative approach allows you to refine prompts gradually. Combining LLMs with reliable tools like pandas for standard tasks reinforces data integrity, and maintaining a copy of the original dataset ensures you can revert if errors occur. Implementing version control can also help track changes and provide a fallback option. With these strategies, you can enhance the reliability and accuracy of data transformations performed by language models.

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [17]:
# Code to rename all the columns in the dataset

# Dictionary for renaming columns
rename_dict = {
    'age': 'Age',
    'bp': 'Blood Pressure',
    'bgr': 'Blood Glucose Random',
    'sg': 'Specific Gravity',
    'al': 'Albumin',
    'su': 'Sugar',
    'hemo': 'Hemoglobin',
    'pcv': 'Packed Cell Volume',
    'wbcc': 'White Blood Cell Count',
    'rbcc': 'Red Blood Cell Count',
    'htn': 'Hypertension',
    'dm': 'Diabetes Mellitus',
    'cad': 'Coronary Artery Disease',
    'appet': 'Appetite',
    'pe': 'Pedal Edema',
    'ane': 'Anemia',
    'Target': 'CKD Status'  # Assuming you have 'Target' for labels
}

# Rename columns in the cleaned data frame
cleaned_combined_df.rename(columns=rename_dict, inplace=True)

# Print the dataset with renamed columns to verify
print("Renamed columns dataset: \n", cleaned_combined_df.head())


Renamed columns dataset: 
          Age  Blood Pressure  Blood Glucose Random        bu        sc  \
6   0.783784            0.00              0.154506  0.111111  0.037037   
7   0.472973            0.50              0.223176  0.098039  0.049383   
8   0.486486            0.25              0.098712  0.143791  0.061728   
9   0.918919            0.00              0.077253  0.261438  0.024691   
11  0.851351            0.25              0.832618  0.503268  0.283951   

         sod       pot  Hemoglobin  Packed Cell Volume  \
6   1.000000  0.137931    0.663366            0.612903   
7   0.600000  0.724138    0.930693            0.580645   
8   0.766667  0.620690    0.881188            0.645161   
9   0.900000  0.275862    0.940594            0.967742   
11  0.333333  0.379310    0.475248            0.451613   

    White Blood Cell Count  ...        pc         pcc          ba  \
6                 0.306452  ...    normal  notpresent  notpresent   
7                 0.387097  ...    normal

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [22]:
# Group by the 'CKD Status' column and count occurrences
target_counts = cleaned_combined_df.groupby('CKD Status').size().reset_index(name='count')

# Print the resulting DataFrame
print("Target counts: \n", target_counts)


Target counts: 
   CKD Status  count
0        ckd     22
1     notckd    111


The provided code groups the `cleaned_combined_df` DataFrame by the `'CKD Status'` column, which indicates whether a patient has chronic kidney disease (CKD) or not. It then counts the occurrences of each unique value in this column, returning a new DataFrame called `target_counts` with two columns: `'CKD Status'`, which contains the unique CKD classifications, and `'count'`, which denotes how many times each classification appears in the dataset. This allows for an easy analysis of the distribution of CKD cases within the dataset.